### PR TITLE
Implement PassGenerator Class

### LINKED ISSUE #372: Implement Class/Node To Find Good Passes
### Description of the task
Implement Class/Node To Find Good Passes

### Outline the task in more detail
We should implement a new ROS node that allows us to generate good passes for the robots to perform. The reasoning behind putting this in a separate node is that this will likely be a costly operation, and so we would like it to be performed in parallel to normal AI operation (in the previous codebase it was on it's own thread). The pass should be computed using gradient descent (implemented in #371) and also by pruning poor passes and generating new ones after some number of iterations of gradient descent.

This node should consume:
- A "World" ROS message representing the current state of the world
- A "Point" ROS message representing the point that the pass should start at
This node should return a "Pass" message (to be defined), containing:
- The point that the pass starts from
- The point that the pass ends from
- The time that the pass should happen at
- The speed of the pass

### Acceptance criteria

- [ ] Aforementioned node is implemented
- [ ] Unit tests are implemented for any classes
- [ ] ROSTests are implemented to test to the overall node functionality

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
Created a `PassGenerator` class that runs gradient descent on a separate thread and allows the user of the class to poll for the best pass found so far. Currently has rough implementations for pass quality and pass generation, those have been pushed off to #383 and #382 in the interests of keeping this focused.

### Resolved Issues
resolves #372

### Length Justification
This is really just one class. I also snuck in a small change where I merged all the `passing` tests together in a group to keep in line with the changes made in #386.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/ai/passing/evaluation.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
+// TODO: start of file comment
+
+#include "ai/passing/evaluation.h"
+#include "util/parameter/dynamic_parameters.h"
+
+double getStaticPositionQuality(Field field, Point position) {
+
+    double positionQuality = 1;
+    double length          = field.length() / 2;
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/evaluation.cpp]: Consider naming these variables `half_field_length` and `half_field_width` for clarity?

File: src/thunderbots/software/ai/passing/evaluation.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        positionQuality =
+                positionQuality / (1 + std::exp(sigmoid_steepness * (position.y() - (width - y_offset))));
+    }
+    else if (position.y() < 0)
+    {
+        positionQuality =
+                positionQuality / (1 + std::exp(sigmoid_steepness * (-position.y() - (width - y_offset))));
+    }
+
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/evaluation.cpp]: As a general note for this function, it would be a good idea to have a negative sigmoid inside the enemy's defense area, since we can't actually receive a ball in that area (The negative sigmoid would cancel out the positive ones in that area. Basically we want the score to drop to 0 inside the enemy defense area). We can pass through it, but the receiver point should not be inside the enemy defense area.

Adding a few sigmoids to cancel out the positive values in that area should do it.

File: src/thunderbots/software/ai/passing/evaluation.h [hunk 3]
```diff
@@ -0,0 +1,18 @@
+// TODO: start of file comment
+
+// TODO: if this file (or the corresponding `.cpp`) gets too big, they may need to be broken up
+#pragma once
+
+#include "ai/world/field.h"
+#include "geom/point.h"
+
+/**
+ * Calculates the static position quality for a given position on a given field
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/evaluation.h]: We should provide a better explanation of what "static position quality" is here.

Eg. "Static position quality prefers good passing points on the field from the perspective of a "real soccer player". For example, passing in front of or towards your own net is less desirable than passing near the enemy's net"

File: src/thunderbots/software/ai/passing/evaluation.h [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * @param field The field on which to calculate the static position quality
+ * @param position The position on the field at which to calculate the quality
+ *
+ * @return A value in [0,1] representing the quality of the given point on the given field
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/evaluation.h]: We should be clear that a higher number is better / represents a more desirable point just to be explicit. This will apply to any other "scoring" evaluation functions.

File: src/thunderbots/software/ai/passing/main.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
+#include <ros/ros.h>
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/main.cpp]: Now that the `PassGenerator` is a threaded object, I don't think this file is necessary anymore?

File: src/thunderbots/software/ai/passing/pass.h [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+         *
+         * @param passer_point The point the pass should start at
+         * @param receiver_point The point the receiver should be at to receive the pass
+         * @param pass_speed_m_per_s The speed of the pass, in meters/second
+         * @param pass_start_time The time that the pass should start at
+         */
+        Pass(Point passer_point, Point receiver_point, double pass_speed_m_per_s, Timestamp pass_start_time);
+
+        /**
+         * Returns a reference to the receiver point
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/pass.h]: The return types for the getters are no longer references. Comments should be updated (applies to the other functions in this file).

File: src/thunderbots/software/ai/passing/pass_generator.cpp [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        // Take ownership of the `in_destructor` flag so we can use it for the conditional
+        // check
+        in_destructor_mutex.lock();
+    }
+}
+
+void PassGenerator::optimizePasses() {
+    // The objective function we minimize in gradient descent to improve each pass
+    // that we're optimizing
+    static const auto objective_function =
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/pass_generator.cpp]: Any reason this is `static`?

File: src/thunderbots/software/ai/passing/pass_generator.cpp [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+            [this](std::array<double, NUM_PARAMS_TO_OPTIMIZE> pass_array){
+                Pass pass = convertArrayToPass(pass_array);
+                return ratePass(pass);
+            };
+
+    // Run gradient descent to optimize the passes to for the requested number
+    // of iterations
+    // TODO: Can we parallize this? Would it make it faster?
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/pass_generator.cpp]: Throwback to `#pragma omp_parallel`?

File: src/thunderbots/software/ai/passing/pass_generator.cpp [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+                objective_function,
+                convertPassToArray(pass),
+                number_of_gradient_descent_steps_per_iter
+        );
+        pass = convertArrayToPass(pass_array);
+    }
+}
+
+void PassGenerator::pruneAndReplacePasses() {
+    // Sort the passes by quality
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/pass_generator.cpp]: We should specify if this is in order of increasing or decreasing quality to be explicit and so people don't have to read the `comparePassQuality` function

File: src/thunderbots/software/ai/passing/pass_generator.cpp [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+                new_passes.begin(), new_passes.end());
+    }
+
+}
+
+void PassGenerator::saveBestPath() {
+    // Take ownership of the best_known_pass for the duration of this function
+    std::lock_guard<std::mutex> best_known_pass_lock(best_known_pass_mutex);
+
+    std::sort(passes_to_optimize.begin(), passes_to_optimize.end(),
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/pass_generator.cpp]: Again, we should specify if we are sorting in increasing or decreasing order.

File: src/thunderbots/software/ai/passing/pass_generator.h [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        // Delete the copy and assignment operators because this class really shouldn't
+        // need them and we don't want to risk doing anything nasty with the internal
+        // threading this class uses
+        PassGenerator & operator=(const PassGenerator&) = delete;
+        PassGenerator(const PassGenerator&) = delete;
+
+        /**
+         * Create a PassGenerator with given parameters
+         *
+         * @param min_reasonable_pass_quality A value in [0,1] representing the minimum
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/pass_generator.h:51]: Should be extra explicit and specify that larger values are "better"

File: src/thunderbots/software/repro.cpp [hunk 12]
```diff
@@ -0,0 +1,18 @@
+#include "ai/passing/pass.h"
```
Comment (MathewMacDougall) [inline src/thunderbots/software/repro.cpp]: I think we solved this issue so this file can be removed?

File: src/thunderbots/software/param_server/README.md [hunk 13]
```diff
@@ -4,11 +4,11 @@ The param server node initializes the dynamic_reconfigure server and calls the `
 
 We currently support `int32_t`, `double`, `bool` and `string`
 ### Registering a parameter to be configured through rqt_reconfigure
-Add the following code in the Params.cfg file found at `param_server/cfg/Params.cfg`
+Add the following code in the ParamArray.cfg file found at `param_server/cfg/ParamArray.cfg`
```
Comment (MathewMacDougall) [inline src/thunderbots/software/param_server/README.md]: Why did the `Params.cfg` filename change here? Accidental refactor?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Overall looking good so far.

File: src/thunderbots/software/ai/passing/evaluation.cpp (hunk 1 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/evaluation.cpp]: Changed

File: src/thunderbots/software/ai/passing/evaluation.cpp (hunk 2 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/evaluation.cpp]: Roughly added that, will do once I get around to adding a circular sigmoid function

File: src/thunderbots/software/ai/passing/evaluation.h (hunk 3 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/evaluation.h]: added that basically verbatiim

File: src/thunderbots/software/ai/passing/evaluation.h (hunk 4 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/evaluation.h]: done

File: src/thunderbots/software/ai/passing/main.cpp (hunk 5 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/main.cpp]: nope, removed

File: src/thunderbots/software/ai/passing/pass.h (hunk 6 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass.h]: donee

File: src/thunderbots/software/ai/passing/pass_generator.cpp (hunk 7 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp]: because it doesn't changed? changed to just `const` to be safe

File: src/thunderbots/software/ai/passing/pass_generator.cpp (hunk 8 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp]: Yeah, I don't think it's worth it here, it doesn't run for long enough. I'll leave a `NOTE` for the future

File: src/thunderbots/software/ai/passing/pass_generator.cpp (hunk 9 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp]: added

File: src/thunderbots/software/ai/passing/pass_generator.cpp (hunk 10 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp]: added

File: src/thunderbots/software/ai/passing/pass_generator.h (hunk 11 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.h:51]: yep, added

File: src/thunderbots/software/repro.cpp (hunk 12 above)
Comment (garethellis0) [reply src/thunderbots/software/repro.cpp]: yup,r emoved

File: src/thunderbots/software/param_server/README.md (hunk 13 above)
Comment (garethellis0) [reply src/thunderbots/software/param_server/README.md]: ah, got aggressive with renaming....... whoops, reverted

**Comment by garethellis0** [comment]
Ok, this should be good to go for a full review now!! GIMME YO COMMENTS

File: src/thunderbots/software/ai/passing/pass_generator.h [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+         * @param passer_point the point we are passing from
+         */
+        void setPasserPoint(Point passer_point);
+
+        /**
+         * Set the target region that we would like to pass to
+         *
+         * @param area The area to pass to
+         */
+        void setTargetRegion(Rectangle area);
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/pass_generator.h]: Should we also have a `clearTargetRegion` function? Right now once we set the target region it will never be `std::nullopt` again and always incorporated in the calculation. The only way to get back to evaluating over the entire field would be to pass the entire field area as a rectangle.

File: src/thunderbots/software/test/ai/passing/pass_generator.cpp [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    std::this_thread::sleep_for(std::chrono::seconds(5));
+
+    std::optional<Pass> pass1 = pass_generator->getBestPassSoFar();
+
+    // Make sure we got some pass
+    ASSERT_TRUE(pass1);
+
+    // Check that the pass is across the half-line towards the enemy goal
+    EXPECT_GE(pass1->receiverPoint().x(), 0.1);
+    // Currently we just generate receiver points at (0,0), so y should be 0
```
Comment (MathewMacDougall) [inline src/thunderbots/software/test/ai/passing/pass_generator.cpp:46]: nit: The method of generating receiver points isn't why y ends up at 0, it's because that gets it closer to the enemy goal.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Just some small nits. Other than the remaining TODOs, all looks good :+1:

File: src/thunderbots/software/ai/passing/pass_generator.cpp [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+        std::thread([this]() { return continuouslyGeneratePasses(); });
+}
+
+void PassGenerator::setWorld(World world)
+{
+    // Take ownership of the world for the duration of this function
+    std::lock_guard<std::mutex> world_lock(world_mutex);
+
+    // Update the world
+    this->world = std::move(world);
```
Comment (JamesPetrie) [inline src/thunderbots/software/ai/passing/pass_generator.cpp:36]: I'm not familiar with the rest of the structure for 'world', but I just want to double check that this won't cause threading issues. In the old AI, world was an object that contained pointers or references to the underlying data structures. So we chose to do a deep copy of the information we needed so we didn't need to worry about things being modified and read at the same time. Is this a non-issue with the new AI / ROS structure?

File: src/thunderbots/software/ai/passing/pass_generator.h [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+         *
+         * @param world
+         */
+        void setWorld(World world);
+
+        /**
+         * Updates the point that we are passing from
+         *
+         * @param passer_point the point we are passing from
+         */
```
Comment (JamesPetrie) [inline src/thunderbots/software/ai/passing/pass_generator.h:70]: Add a warning about the old pass points being reset when this function is called. I think it does make sense to do that, but if someone isn't aware they might accidentally call setPasserPoint() once per tick and not know why the optimization isn't converging.

File: src/thunderbots/software/ai/passing/pass_generator.cpp [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+                passes_to_optimize.end());
+        }
+
+        // Append our newly generated passes to replace the passes we just removed
+        passes_to_optimize.insert(passes_to_optimize.end(), new_passes.begin(),
+                                  new_passes.end());
+    }
+}
+
+void PassGenerator::saveBestPath()
```
Comment (JamesPetrie) [inline src/thunderbots/software/ai/passing/pass_generator.cpp]: Typo: should be Pass not Path

File: src/thunderbots/software/ai/passing/pass_generator.cpp [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+        if (num_passes_to_keep_after_pruning.value() < passes_to_optimize.size())
+        {
+            passes_to_optimize.erase(
+                passes_to_optimize.begin() + num_passes_to_keep_after_pruning.value(),
+                passes_to_optimize.end());
+        }
+
+        // Append our newly generated passes to replace the passes we just removed
+        passes_to_optimize.insert(passes_to_optimize.end(), new_passes.begin(),
+                                  new_passes.end());
```
Comment (JamesPetrie) [inline src/thunderbots/software/ai/passing/pass_generator.cpp:174]: With this implementation passes_to_optimize.size() doesn't necessarily equal num_passes_to_optimize. If this is intentional, maybe rename to specify it is the maximum number of passes to optimize.

File: src/thunderbots/software/ai/passing/pass_generator.cpp [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+    while (!in_destructor)
+    {
+        // Give up ownership of the in_destructor flag now that we're done the
+        // conditional check
+        in_destructor_mutex.unlock();
+
+        optimizePasses();
+        pruneAndReplacePasses();
+        saveBestPath();
+
```
Comment (JamesPetrie) [inline src/thunderbots/software/ai/passing/pass_generator.cpp:99]: In the old code there was logic to start from scratch if the best quality was below a certain threshold. This was useful because sometimes the pass points would behave badly when the quality was really low, so that even once the game state had improved those points were still very far from optimal

File: src/thunderbots/software/ai/passing/pass_generator.cpp [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+    // Run gradient descent to optimize the passes to for the requested number
+    // of iterations
+    // NOTE: Parallelizing this `for` loop would probably be safe and potentially more
+    //       performant
+    for (Pass& pass : passes_to_optimize)
+    {
+        auto pass_array =
+            optimizer.maximize(objective_function, convertPassToArray(pass),
+                               number_of_gradient_descent_steps_per_iter.value());
```
Comment (JamesPetrie) [inline src/thunderbots/software/ai/passing/pass_generator.cpp:128]: Maybe add a TODO to test the performance of some different sets of hyperparameters for the optimization being used

File: src/thunderbots/software/ai/passing/pass_generator.cpp [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+{
+    return {pass.receiverPoint().x(), pass.receiverPoint().y(), pass.speed(),
+            pass.startTime().getSeconds()};
+}
+
+Pass PassGenerator::convertArrayToPass(
+    std::array<double, PassGenerator::NUM_PARAMS_TO_OPTIMIZE> array)
+{
+    // Take ownership of the passer_point for the duration of this function
+    std::lock_guard<std::mutex> passer_point_lock(passer_point_mutex);
```
Comment (JamesPetrie) [inline src/thunderbots/software/ai/passing/pass_generator.cpp:268]: don't need the passer mutex here

Comment (JamesPetrie) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:268]: Never mind- realized that you actually do

File: src/thunderbots/software/ai/passing/pass_generator.h [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+        // The number of parameters (representing a pass) that we optimize
+        // (pass_start_x, pass_start_y, pass_speed, pass_start_time)
+        static const int NUM_PARAMS_TO_OPTIMIZE = 4;
+
+
+        // Weights used to normalize the parameters that we pass to GradientDescent
+        // (see the GradientDescent documentation for details)
+        static constexpr double PASS_SPACE_WEIGHT                          = 0.01;
+        static constexpr double PASS_TIME_WEIGHT                           = 1;
+        static constexpr double PASS_SPEED_WEIGHT                          = 1;
```
Comment (JamesPetrie) [inline src/thunderbots/software/ai/passing/pass_generator.h:111]: I think the ratio we used before was 1:2:6, why the change? The ratio we were using was picked using the heuristic of normalizing the range available to each parameter. So if the spatial range is [-4,4], time is [0,3], and speed is [2,6] then the weights would be 1:(8/3):2. There are probably better ways to do this though

File: src/thunderbots/software/ai/passing/pass_generator.cpp [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+
+std::optional<Pass> PassGenerator::getBestPassSoFar()
+{
+    // Take ownership of the best_known_pass for the duration of this function
+    std::lock_guard<std::mutex> best_known_pass_lock(best_known_pass_mutex);
+
+    return best_known_pass;
+}
+
+void PassGenerator::setTargetRegion(Rectangle area)
```
Comment (JamesPetrie) [inline src/thunderbots/software/ai/passing/pass_generator.cpp]: Good idea

**Comment by JamesPetrie** [comment]
Looks good overall. Much cleaner structure than before

File: src/thunderbots/software/ai/passing/pass_generator.h (hunk 14 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.h]: Changed this take an `std::optional<Rectangle>`, so the user can specify no region by giving it `std::nullopt`. Changed the javadoc comment to reflect this.

File: src/thunderbots/software/test/ai/passing/pass_generator.cpp (hunk 15 above)
Comment (garethellis0) [reply src/thunderbots/software/test/ai/passing/pass_generator.cpp:46]: updated comment

File: src/thunderbots/software/ai/passing/pass_generator.cpp (hunk 16 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:36]: @MathewMacDougall to confirm, but I don't believe we hold anything by pointer in the World object or it's members

File: src/thunderbots/software/ai/passing/pass_generator.h (hunk 17 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.h:70]: Added

File: src/thunderbots/software/ai/passing/pass_generator.cpp (hunk 18 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp]: changed

File: src/thunderbots/software/ai/passing/pass_generator.cpp (hunk 19 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:174]: Sorry, which statement are you referring to here?

File: src/thunderbots/software/ai/passing/pass_generator.cpp (hunk 20 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:99]: Hmm, this is a great point. I'm going to make a note if it on the issue to implement `generatePasses` because I think it fits better in that context? (I'll be sure to tag you for review on that as well)

File: src/thunderbots/software/ai/passing/pass_generator.cpp (hunk 21 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:128]: Great point, we should definitely invest considerable effort in tuning, could significantly improve results. Opened https://github.com/UBC-Thunderbots/Software/issues/388 to tune all the various hyperparameters after we've finished implementation.

File: src/thunderbots/software/ai/passing/pass_generator.h (hunk 23 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.h:111]: No particular reason, just had some random values that worked somewhat well for convergence over the static field. These will definitely get changed as part of #388 though.

**Comment by garethellis0** [comment]
Comments addressed, ready for re-review!

File: src/thunderbots/software/ai/passing/pass_generator.cpp (hunk 19 above)
Comment (JamesPetrie) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:174]: The number of generated passes is always num_passes_to_optimize.value() - num_passes_to_keep_after_pruning.value()

But the number of passes in passes_to_optimize may be less than num_passes_to_keep_after_pruning.value() because of the merging

So merged + generated <= num_passes_to_optimize

File: src/thunderbots/software/ai/passing/pass_generator.cpp (hunk 16 above)
Comment (JamesPetrie) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:36]: So will the world object used by the passing thread be completely unchanged except when setWorld() is called? And if that is the case now can it be guaranteed for future versions of the World class?

I don't see any immediate issues if it is changing in parallel, but this is the sort of thing that could lead to those super rare threading bugs that are impossible to debug

Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:36]: Yeah, that's correct. Most of the system makes this assumption, it was a specific decision made early on to make things easier to reason about.

File: src/thunderbots/software/ai/passing/pass_generator.cpp (hunk 19 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:174]: Ah, great catch. Changed to:
```
        // Remove the worst paths
        if (num_passes_to_keep_after_pruning.value() < passes_to_optimize.size())
        {
            passes_to_optimize.erase(
                passes_to_optimize.begin() + num_passes_to_keep_after_pruning.value(),
                passes_to_optimize.end());
        }

        // Generate new passes to replace the ones we just removed
        std::vector<Pass> new_passes = generatePasses(
            num_passes_to_optimize.value() - passes_to_optimize.size());

        // Append our newly generated passes to replace the passes we just removed
        passes_to_optimize.insert(passes_to_optimize.end(), new_passes.begin(),
                                  new_passes.end());
```

Comment (JamesPetrie) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:174]: Looks good

File: src/thunderbots/software/ai/passing/pass_generator.cpp (hunk 16 above)
Comment (JamesPetrie) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:36]: I definitely agree with that design decision. Is there a way to guarantee that the copied world object has no links to the original? Either through class design, tests, or just a bunch of comments warning about how World and the objects it owns can be designed

The scenario I'm concerned about is if 2 years from now someone accidentally adds an array to one of the objects owned by world, and then each time world gets copied both objects actually point to the same memory

Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:36]: Great point. Added the following comment to the javadoc comment for the `World` class:
```
 * WARNING: This class should _never_ hold any data that is pointed to anywhere else. This
 * means no raw pointers, no shared pointers, no shared references. This is because in
 * some cases we copy World's over multiple threads where having a shared member
 * between multiple instances of a World on multiple threads could result in
 * data corruption that would lead to absurdly hard to track bugs. YOU HAVE BEEN WARNED.
```

Comment (MathewMacDougall) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:36]: I can confirm that the `World` currently doesn't hold references/pointers to any of its members, and this is not planned for the future. I agree with the comment that was added, and think we should be safe from threading issues.

Comment (JamesPetrie) [reply src/thunderbots/software/ai/passing/pass_generator.cpp:36]: Looks good

File: src/thunderbots/software/param_server/cfg/Params.cfg [hunk 25]
```diff
@@ -50,6 +50,12 @@ navigator_params = gen.add_group("Passing")
 navigator_params.add("static_field_position_quality_x_offset", double_t, 0, "", 0.3, 0, 1)
 navigator_params.add("static_field_position_quality_y_offset", double_t, 0, "", 0.3, 0, 1)
 navigator_params.add("static_field_position_quality_friendly_goal_distance_weight", double_t, 0, "", 0.3, 0, 1)
+navigator_params.add("num_passes_to_optimize", int_t, 0, "", 50, 0, 1000)
+navigator_params.add("num_passes_to_keep_after_pruning", int_t, 0, "", 10, 0, 1000)
+navigator_params.add("number_of_gradient_descent_steps_per_iter", int_t, 0, "", 20, 0, 1000)
+navigator_params.add("pass_equality_max_position_difference_meters", double_t, 0, "", 0.1, 0, 4)
+navigator_params.add("pass_equality_max_start_time_difference_seconds", double_t, 0, "", 0.5, 0, 3)
+navigator_params.add("pass_equality_max_speed_difference_meters_per_second", double_t, 0, "", 0.3, 0, 4)
```
Comment (JamesPetrie) [inline src/thunderbots/software/param_server/cfg/Params.cfg:58]: Should have asked before, but are these dynamic parameters thread safe?

Comment (garethellis0) [reply src/thunderbots/software/param_server/cfg/Params.cfg:58]: @akhilveeraghanta - can you confirm?

Comment (garethellis0) [reply src/thunderbots/software/param_server/cfg/Params.cfg:58]: I think we should be good here because these are never updated from within ai, just through external means.

Comment (MathewMacDougall) [reply src/thunderbots/software/param_server/cfg/Params.cfg:58]: Just my 2c, but we should be fine. @garethellis0 is correct we only ever read the parameters in the AI (AI never sets them), and ROS is responsible for safely updating them when they change externally.

Comment (JamesPetrie) [reply src/thunderbots/software/param_server/cfg/Params.cfg:58]: Sounds good

**Review by JamesPetrie — DISMISSED** [review]
Looks good to me, nice work

**Review by MathewMacDougall — DISMISSED** [review]

File: src/thunderbots/software/param_server/cfg/Params.cfg (hunk 25 above)
Comment (akhilveeraghanta) [reply src/thunderbots/software/param_server/cfg/Params.cfg:58]: The ROS Param server api itself is _not_ thread safe, but we don't directly set the parameters managed by the DynamicReconfigure server through `ros::param::set`, we do it through the UI which makes a service call to the reconfigure server, which is thread safe. So we should be fine as long as we use the gui, if we start setting them directly we can run into problems.

**Review by MathewMacDougall — APPROVED** [review]
